.github/actions/vcpkg-cache-action/main.py

```diff
@@ -0,0 +1,197 @@
+
+import aiohttp
+import aiofiles
+
+LOCALHOST = "127.0.0.1"
+
+logging.basicConfig()
+logger = logging.getLogger()
+
+
+class ProxyStartError(RuntimeError):
+    pass
+
+
+def find_free_port() -> int:
+    with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as s:
+        s.bind((LOCALHOST, 0))
+        return s.getsockname()[1]
+
+
+async def wait_until_healthy(proc: subprocess.Popen, port: int):
+    url = f"http://{LOCALHOST}:{port}/health"
+    async with aiohttp.ClientSession() as session:
+        for x in range(10):
+            if proc.poll() is not None:
+                raise ProxyStartError(
+                    f"Proxy process (PID {proc.pid}) exited unexpectedly with code {proc.returncode}"
+                )
+
+            try:
+                async with session.get(url, timeout=0.5) as response:
+                    if response.ok:
+                        return
+            except Exception:
+                pass
+
+            await asyncio.sleep(0.5)
+    raise ProxyStartError("Proxy did not become healthy in time")
+
+
+async def tail(log_path: str):
+    async with aiofiles.open(log_path, mode="rb") as log_file:
+        while True:
+            buffer = await log_file.read(128)
+            if buffer == b"":
+                await asyncio.sleep(0.1)
+            await aiofiles.stderr_bytes.write(buffer)
+            await aiofiles.stderr.flush()
+
+
+def isGitHubActions() -> bool:
+    return os.getenv("GITHUB_ACTIONS") == "true"
+
+
+def isPost() -> bool:
+    return get_state("isPost") == "true"
+
+
+def get_state(key: str) -> str | None:
+    return os.getenv(f"STATE_{key}")
+
+
+async def set_state(key: str, value: str) -> None:
+    if isGitHubActions():
+        async with aiofiles.open(os.getenv("GITHUB_STATE"), mode="a") as state_file:
+            await state_file.write(f"{key}={value}\n")
+    else:
+        logger.info(f"STATE {key}={value}")
+
+
+async def set_env(key: str, value: str) -> None:
+    if isGitHubActions():
+        async with aiofiles.open(os.getenv("GITHUB_ENV"), mode="a") as env_file:
+            await env_file.write(f"{key}={value}\n")
+    else:
+        logger.info(f"ENV {key}={value}")
+
+
+def proc_exists(pid: int) -> bool:
+    try:
+        os.kill(pid, 0)
+    except OSError:
+        return False
+    else:
+        return True
+
+
+async def wait_for_proc(pid: int) -> None:
+    while proc_exists(pid):
+        await asyncio.sleep(0.1)
+
+
+async def start() -> None:
+    port = find_free_port()
+
+    cmd = [
+        sys.executable,
+        "-m",
+        "uvicorn",
+        "--port",
+        str(port),
+        "--use-colors",
+        "proxy:app",
+    ]
+
+    log_fd, log_path = tempfile.mkstemp()
+    tail_task = asyncio.create_task(tail(log_path))
+    proc: subprocess.Popen | None = None
+
+    try:
+        # asyncio.subprocess does not allow us to create detached processes on
+        # Windows so we use subprocess.Popen here
+        proc = subprocess.Popen(
```

Comment:
Should `creationflags=subprocess.DETACHED_PROCESS` be used on Windows? I guess it obviously works as-is so it's not needed, but maybe that would solve the problem with asyncio.subprocess?

Reply:
I did try that with `asyncio.subprocess`, but it didn't work. I even tried adding more flags, like `CREATE_NEW_PROCESS_GROUP` and `CREATE_BREAKAWAY_FROM_JOB`, but it didn't work either. The issue seems to be with how the asyncio event loop closes the child process on exit, but only on Windows. I am not entirely sure!

Reply:
So the loop was already a `ProactorEventLoop` on Windows (otherwise launching a subprocess would not work at all), and in my testing it looks like it is indeed the one closing the child process, no matter which creation flags I pass. I'm giving up 🤷